### PR TITLE
DrivAerML: learned error correction head (self-correcting predictions)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The remaining DM error at 3.833% is **systematic spatial bias** — the model consistently over/under-predicts in specific geometric regions (sharp junctions, underbody, A-pillar). This means the prediction errors have learnable spatial structure. We hypothesize that a **secondary correction head** trained to predict and subtract the model's own errors can exploit this structure to reduce the final prediction error.

The approach is inspired by **gradient boosting** and **iterative refinement**: the primary head makes a coarse prediction, then a correction head learns the residual error pattern from the same latent representation and adds a correction term. During training, the correction head's target is the primary head's detached prediction error (target - primary_pred). At inference, the final prediction is primary_pred + correction.

This is fundamentally different from residual prediction (dead for DM — that predicted corrections to freestream). Here, the correction head operates on the model's OWN intermediate representation and corrects the model's OWN errors, not corrections to a physics prior. The primary head provides the scaffold; the correction head refines it.

**Key design choice:** The correction head must be lightweight (1-2 linear layers) to avoid capacity competition with the primary head. And the primary head's prediction must be DETACHED when computing the correction target — otherwise the gradients would fight.

### Implementation

1. **Add CLI flag:**
   ```python
   parser.add_argument('--error-correction-head', action='store_true', default=False,
                       help='Add a secondary head that learns to correct primary prediction errors')
   parser.add_argument('--correction-loss-weight', type=float, default=0.1,
                       help='Weight for the correction head loss')
   ```

2. **Add correction head to the model** (in `__init__`):
   ```python
   if args.error_correction_head:
       # Lightweight MLP: takes latent features, outputs correction per channel
       self.correction_head = nn.Sequential(
           nn.Linear(hidden_dim, hidden_dim // 4),
           nn.GELU(),
           nn.Linear(hidden_dim // 4, output_dim)
       )
   ```

3. **Modify the forward pass** to produce both primary and correction predictions:
   ```python
   primary_pred = self.output_head(latent_features)  # existing prediction
   
   if hasattr(self, 'correction_head'):
       correction = self.correction_head(latent_features)
       final_pred = primary_pred + correction
   else:
       final_pred = primary_pred
   ```

4. **Compute the correction loss** (in the training loss computation):
   ```python
   # Primary loss (on final corrected prediction)
   primary_loss = mse_loss(final_pred, target)
   
   if args.error_correction_head:
       # Correction target: the primary head's error (detached!)
       with torch.no_grad():
           primary_error = target - primary_pred.detach()
       correction_loss = mse_loss(correction, primary_error)
       total_loss = primary_loss + args.correction_loss_weight * correction_loss
   else:
       total_loss = primary_loss
   ```

   **Critical:** `primary_pred.detach()` in the correction target — this prevents the correction loss from back-propagating through the primary head, which would create conflicting gradient signals.

5. **At evaluation/test time:** use `final_pred = primary_pred + correction` as the prediction. No changes needed since the forward pass already produces this.

### Trials

**Smoke test first (3 epochs)** — verify shapes, both losses finite, correction head producing non-zero output.

**Trial 1 — correction weight=0.1 (moderate correction signal):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --error-correction-head \
  --correction-loss-weight 0.1 \
  --wandb_group dm-error-correction-head
```

**Trial 2 — correction weight=0.5 (stronger correction signal):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --error-correction-head \
  --correction-loss-weight 0.5 \
  --wandb_group dm-error-correction-head
```

Log the correction head's loss separately (correction_loss) — if it collapses to near-zero quickly, the primary head is already doing all the work and the correction head adds nothing. If it stays elevated, the correction head is actively learning residual patterns.

Report `val_primary/surface_rel_l2_pct` for both trials. Target: beat **3.833%**.

## Baseline

Current best DrivAerML metrics (PR #3072, eren/dm-ema-9995-gc05):

| Metric | Value | Epoch |
|--------|-------|-------|
| val_primary/surface_rel_l2_pct | **3.833%** | 511 |
| test_primary/surface_rel_l2_pct | **4.685%** | 511 |

- W&B run: `ncl1dh88`
- Config: 4L/512d/8H, AdamW lr=5e-4, T_max=30, gc=0.5, EMA=0.9995, no WD, Fourier

Reproduce:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT emma:

## Results

### Trial 1 — correction_loss_weight=0.1

| Metric | Value |
|--------|-------|
| **val_primary/surface_rel_l2_pct** | **9.169%** (epoch 79) |
| test surface_rel_l2_pct (best ckpt) | 12.864% |
| correction_loss (final) | 0.71 (diverged) |
| Total epochs | 225 (diverged at ~epoch 80, never recovered) |
| W&B run | `vvh651d8` |

**Status: DIVERGED.** Training became unstable around epoch 80 — grad norms exploded to Infinity and the model never recovered. The lighter correction weight (0.1) appeared insufficient to stabilize the correction head's interaction with the primary loss.

**Command:**
```bash
CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --error-correction-head --correction-loss-weight 0.1 \
  --agent emma --wandb-name "emma/dm-error-correction-w0.1" --wandb-group dm-error-correction-head
```

### Trial 2 — correction_loss_weight=0.5

| Metric | Value |
|--------|-------|
| **val_primary/surface_rel_l2_pct** | **4.700%** (epoch 227) |
| test surface_rel_l2_pct (best ckpt) | 5.535% |
| test surface_rel_l2_pct (final) | 5.497% |
| correction_loss (final) | 0.006 |
| Total epochs | 228 (hit 180-min timeout) |
| W&B run | `riq65o1p` |

**Command:**
Same as Trial 1 but with `--correction-loss-weight 0.5`.

### Baseline comparison

| | Baseline (ncl1dh88) | Trial 1 (w=0.1) | Trial 2 (w=0.5) |
|--|--|--|--|
| val surface_rel_l2_pct | **3.833%** (ep 511) | 9.169% (ep 79, diverged) | 4.700% (ep 227) |
| test surface_rel_l2_pct | 4.685% | 12.864% | 5.535% |
| External target | <3.71% (AB-UPT) | — | — |

## What happened

**Neither trial beat the baseline, but Trial 2 shows promise.**

1. **Trial 1 (w=0.1) diverged.** Around epoch 80, grad norms exploded to Infinity and training collapsed. The lighter correction weight didn't provide enough stabilization. The correction head's gradients likely conflicted with the primary head's gradients at a cosine cycle restart, and without sufficient correction loss magnitude, the optimizer couldn't recover.

2. **Trial 2 (w=0.5) converged stably** to 4.700% val in 227 epochs. This is ~0.87pp above the 3.833% baseline, but the baseline ran to epoch 511 and Trial 2 was cut short at 228 epochs by the 180-min timeout. The convergence trajectory at epoch 227 was still improving.

3. **The correction head's loss stayed active** throughout training for Trial 2 (0.006 at end), indicating it was continuously learning residual error patterns rather than collapsing — a positive sign for the hypothesis.

4. **The epoch-matched comparison is unfavorable**: at ~200 epochs, the previous baseline experiments (from PR #3258 aux normals) reached ~4.66-4.72%, and Trial 2 here is at 4.70% — essentially identical. The correction head isn't adding measurable improvement at this epoch count.

## Implementation notes

- The correction head takes `[latent_features, primary_pred.detach()]` as input (concat), following research recommendations that conditioning on the primary prediction helps the correction head learn where errors occur.
- Zero-initialized final layer ensures corrections start from zero and grow gradually.
- The correction is applied at both train and eval time: `final_pred = primary_pred + correction`.
- Correction loss target: `(target - primary_pred.detach())`, trained with MSE.

## Suggested follow-ups

1. **Extended training** — Trial 2 needs the full 500+ epoch budget to determine if the correction head provides a real advantage over the baseline at convergence.
2. **Lower correction weight with warmup** — A correction weight schedule that starts at 0 and ramps up after the primary head has converged could prevent the instability seen in Trial 1 while still providing meaningful correction signal later.
3. **Tanh-bounded correction** — Add `alpha * tanh(correction)` with learned alpha initialized to 0, preventing the correction from growing too large and destabilizing training.
4. **Stronger grad clip** — The divergence in Trial 1 might be prevented with a tighter grad clip (0.3 instead of 0.5).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3269 because ADVISOR: Closing — error correction head does not improve DM convergence rate. Trial 2 (w=0.5) at 4.700% val after 227 epochs is epoch-matched identical to baseline (~4.66-4.72% at ep200). The correction head adds architectural complexity and computational overhead without measurable benefit. Trial 1 (w=0.1) diverged at ep80 — the correction-primary gradient interaction is fragile at cosine restart peaks. The correction loss staying active (0.006) shows the head learns something, but it's not translating to primary metric improvement. GPU time better spent on champion config variations. Emma reassigned..
